packages/flutter/lib/src/widgets/autofill.dart

```diff
@@ -225,3 +228,142 @@ class _AutofillScope extends InheritedWidget {
+  /// [AutofillContextLifecycleAction] widget that encloses the given [context],
+  /// or the default configuration [AutofillContextLifecycleAction.globalLifecycleDelegate]
+  /// if the given [context] does not contain an [AutofillContextLifecycleDelegate].
+  static AutofillContextLifecycleDelegate of(BuildContext context) {
+    return context?.dependOnInheritedWidgetOfExactType<AutofillContextLifecycleAction>()?.lifecycleDelegate ?? globalLifecycleDelegate;
+  }
+
+  // Changing delegates shouldn't make anything other than the widget itself
+  // rebuild.
+  @override
+  bool updateShouldNotify(covariant AutofillContextLifecycleAction oldWidget) => false;
+}
+
+/// Configures whether the current autofill context should be saved or discarded
+/// , or no action should be taken at all, when [Navigator]s and [Form]s
+/// configured by this [AutofillContextLifecycleDelegate] observes certain types
+/// of events.
+///
+/// Typically retrieved using [AutofillContextLifecycleAction.of].
+///
+/// The default implementation signals the platform to save the current autofill
+/// context (before discarding it), when any of the affected [Form]s is saved,
+/// or any of the affected [Navigator]s pushes/pops opaque [ModalRoute], or the
+/// current route is replaced with an opaque [ModalRoute].
+class AutofillContextLifecycleDelegate {
+  /// Creates a delegate object that can be used to clean up the current
+  /// autofill context at proper times, for example, when [Route]s change or
+  /// when [Form]s are saved.
+  const AutofillContextLifecycleDelegate();
+
+  /// The action needs to be taken after [FormState.save] is called on
+  /// [savedForm].
+  ///
+  /// The default implementation saves the current autofill context.
+  void onFormSave(FormState savedForm) {
+    TextInput.finishAutofillContext(shouldSave: true);
+  }
+
+  /// The action to take when any [Navigator] in the app pushed [route], on top
+  /// of the previous active route [previousRoute].
+  ///
+  /// {@macro flutter.widgets.navigatorObserver.didPush}
+  ///
+  /// The default implementation saves the current autofill context if [route]
+  /// is an opaque [ModalRoute].
+  void didPush(Route<dynamic> route, Route<dynamic> previousRoute) {
+    // TODO(LongCatIsLooong): Ideally we want to set the current context aside
+    // and start a new autofill context for the new route we pushed. This
+    // doesn't seem to be possible on Android currently.
+    if (route is ModalRoute && route.opaque)
+      TextInput.finishAutofillContext(shouldSave: true);
+  }
+
+  /// The action to take when a [Navigator] in the app popped [route].
+  ///
+  /// {@macro flutter.widgets.navigatorObserver.didPop}
+  ///
+  /// The default implementation saves the current autofill context if [route]
+  /// is an opaque [ModalRoute].
+  void didPop(Route<dynamic> route, Route<dynamic> previousRoute) {
+    if (route is ModalRoute && route.opaque)
```

Comment:
Why is this the condition for saving? What if I have my form an an Alert box or something like that?